### ISSUE TITLE
ELPA

### ISSUE THREAD

**Issue by naphthalene** [yoshiki/yaml-mode#44]
It would be nice to be able to pull this from the main package repositories (Emacs 24+)

**Comment by wasamasa** [comment]
I agree.  The main difficulty with adding a package to GNU ELPA is convincing anyone who has contributed non-trivial (read: more than five lines of diff) amounts of code to sign the FSF copyright assignment form.  I have already done this, however it's unclear whether any other contributors did or are willing to.  Due to this complication people only bother with this for either popular packages with a great benefit or new ones that don't have a complicated history.

Ping <strike>@yoshiki</strike> @antalk2 <strike>@mmckinst</strike> <strike>@timcharper</strike> <strike>@SamB</strike> <strike>@Wilfred</strike> <strike>@wentasah</strike> <strike>@llasram</strike> <strike>@Kaali</strike> <strike>@ts4z</strike>.

**Comment by antalk2** [comment]
> It would be nice to be able to pull this from the main package repositories (Emacs 24+)
I agree, too.
> to sign the FSF copyright assignment form.
I would. How can I do that?

**Comment by Wilfred** [comment]
I've already done FSF assignment, FWIW. I know there's a FSF machine that has a list of people who have already assigned: https://github.com/abo-abo/swiper/issues/577#issuecomment-231952793

To assign copyright for your email contributions, just email assign@gnu.org and tell them you want to assign copyright for your copyright contributions. They'll tell you what to do.

**Comment by timcharper** [comment]
Emailed.

**Comment by wentasah** [comment]
I also wrote them.

(For others: They seem to need the information described in http://git.savannah.gnu.org/gitweb/?p=gnulib.git;a=tree;f=doc/Copyright;hb=HEAD)

**Comment by mmckinst** [comment]
There's 4 of them listed in there. Which of those am I supposed to fill out and email them?

**Comment by wasamasa** [comment]
You're not supposed to fill one out upfront.  Write an email to the above address and wait for further instructions.

**Comment by mmckinst** [comment]
OK. I've emailed them. FWIW my contribution was pretty trivial since I just updated the FSF address and added a copy of the GPLv2 but will assign copyright regardless.

**Comment by yoshiki** [comment]
I agree too.
I would send email to assign copyright.

**Comment by naphthalene** [comment]
Any news on this?

**Comment by wentasah** [comment]
I wait for my employer to sign the paper.

**Comment by yoshiki** [comment]
My assignment/disclaimer process with the FSF is complete.
Any update on this?

**Comment by wasamasa** [comment]
I've updated the initial list.  The thing is that inclusion into GNU ELPA can only proceed once every non-trivial submission had its author tracked down and their assignment has been done.

**Comment by wentasah** [comment]
My assignment process is complete as well.

**Comment by mmckinst** [comment]
I just finally mailed mine today, not sure what the turnaround time is for snail mail.

There's still the issue of the code antalk2 submitted...

Once the copyright is all assigned to the FSF, I wonder if this could all just be moved to the emacs project so yaml-mode comes direct from emacs?

**Comment by mmckinst** [comment]
My copyright assignment to the FSF has been completed.

**Comment by wasamasa** [comment]
@mmckinst Thank you!  Regarding your previous question, it's become unusual to include more into the Emacs sources, save for a few very convincing cases (like xref).  It's likelier that with this, yaml-mode could be included into GNU ELPA.

**Comment by naphthalene** [comment]
Looks like @SamB is left then? Thank you all for taking the time to do this! :+1:

**Comment by Alhadis** [comment]
Any update? I tried installing `yaml-mode` using `M-x package-install` and I got `[No match]`. :sad:

**Comment by wasamasa** [comment]
This cannot progress until all non-trivial contributors have handed in copyright assignments.  FWIW, this includes @antalk2 and Marshall T. Vandegrift who I have yet to find on the internets.

**Comment by llasram** [comment]
Marshall T. Vandegrift -- now Bockrath for some extra confusion -- is @llasram.  As requested by e-mail, I've written to assign@gnu.org, and will comment again after completing the process.

**Comment by wasamasa** [comment]
Ping @antalk2 @timcharper @llasram @Kaali.

**Comment by Kaali** [comment]
I've sent the request. Now waiting for forms.

**Comment by rchar01** [comment]
What is the status of this thread?

**Comment by wasamasa** [comment]
I didn't receive any updates outside this thread, so it's unchanged.

**Comment by Kaali** [comment]
My bad. I asked for some clarifications to the assignment contract, and FSF never got back to me. I'll ping them.

**Comment by Kaali** [comment]
Took a while. Got the response today, the process is complete.

**Comment by dgutov** [comment]
@timcharper and @SamB seem to both have their assignments on file. Samuel even from way before (2009).

**Comment by dgutov** [comment]
@llasram doesn't. But I can't find the commits authored by him either.

That leaves @antalk2, I think. But it could be argued that comments to test cases are not part of the work (the program). We can ask.

**Comment by wasamasa** [comment]
@llasram is mentioned in the copyright lines, presumably for a reason. Maybe his bulk contributions didn't survive some VCS conversion? I've edited the other post accordingly.

**Comment by antalk2** [comment]
https://github.com/yoshiki/yaml-mode/issues/44#issuecomment-246765665

You can take my contribution on an AS IS basis, no warranty, no promise of indemnification, 
100% at your own risk.

The FSF proposal I saw appears to me an overreach:
give everything, including shielding FSF from any possible negative consequences,
without even a promise to try or even seriously consider including yaml-mode in emacs or ELPA
in return.

**Comment by dgutov** [comment]
> is mentioned in the copyright lines, presumably for a reason. Maybe his bulk contributions didn't survive some VCS conversion?

Hmm. Actually, it seems like @yoshiki just committed under him own name. `Changes` in the meantime says (and there's a similar line in the commentary):

> Patch from Marshall T. Vandegrift. Added lots of features.

So I guess we shouldn't go forward without him.

> What about @ts4z?

One patch, five lines. Below the limit.

**Comment by dgutov** [comment]
> You can take my contribution on an AS IS basis, no warranty, no promise of indemnification,
100% at your own risk.

In principle, that doesn't work. We could delete it and rewrite it, though, if that turns out to be necessary (or just delete). But your contribution is not big enough to make a fuss about, I think (hope).

> give everything, including shielding FSF from any possible negative consequences

Shielding - yes: if it turns out that you (or someone) committed a patch of code owned by some corporation, and the corporation sues, FSF shouldn't be the main target.

Nothing like this ever happened, though. To my knowledge.

> without even a promise to try or even seriously consider including yaml-mode in emacs or ELPA
in return

What kind of promise do you want? We will include it, as soon as all assignments are done. Why wouldn't we?

**Comment by antalk2** [comment]
> > You can take my contribution on an AS IS basis, no warranty, no promise of indemnification,
> > 100% at your own risk.

> In principle, that doesn't work. We could delete it and rewrite it, though, if that turns out to be 
> necessary (or just delete). 
Please, do whatever os necessary in practice.

> But your contribution is not big enough to make a fuss about, I think (hope).
I hoped that too. Surprised that four years later it came up again.

> give everything, including shielding FSF from any possible negative consequences
> Shielding - yes: if it turns out that you (or someone) committed a patch of code owned by some 
> corporation, and the corporation sues, FSF shouldn't be the main target.

As advertised, FSF needs ownership to have better chance in their legal battles.
They (I assumed) have the resources to engage in shuch things, and I assumed
that in exchange for ownership they provide legal representation. However
their actual proprosal says otherwise: it says the contributors should shield them.
Nor does it promise to enforce any licences, not event the intent.

> > without even a promise to try or even seriously consider including yaml-mode in emacs or ELPA
> > in return

> What kind of promise do you want? We will include it, as soon as all assignments are done.
> Why wouldn't we?

Ami I correct to assume that you (dgutov) represent the FSF?

Well, an informal "Why wouldn't we?" appears much weaker than
including such an intent in the legal paper FSF sends out to contributors.
For the latter basically says: we accept the rights, but not any reponsibilities.
I presume in any case signed papers count more than "Why wouldn't we?"
on a mailing list. So why does the FSF offer not include any reference
to these concerns, not even on a best effort basis?

**Comment by dgutov** [comment]
> As advertised, FSF needs ownership to have better chance in their legal battles.
They (I assumed) have the resources to engage in shuch things

Yes, but if it turns out the suit has merit, they need to be able to protect themselves. Otherwise a malicious actor could get some code (owned by his friend, for instance) into an FSF project, then the friend sues, wins the court case, and they split the money.

> Nor does it promise to enforce any licences, not event the intent.

I'm not sure how such promise would look in legal writing. But you can look at FSF's mission, their projects, and overall track record.

> Ami I correct to assume that you (dgutov) represent the FSF?

I'm just one of the Emacs developers. You can browse my projects.

> So why does the FSF offer not include any reference to these concerns, not even on a best effort basis?

I'm sure there are multiple explanations around the internet. The above is just my understanding of the situation. If you have doubts and questions, by the way, you can also ask them over email to assing@gnu.org.

**Comment by antalk2** [comment]
>> So why does the FSF offer not include any reference to these concerns, 
>> not even on a best effort basis?

> But you can look at FSF's mission, their projects, and overall track record.
> I'm sure there are multiple explanations around the internet.

My worry is: any amount of these outside the agreement is probably considered irrelevant
by lawyers. On the other hand, the agreement does not even mention
the reason for contributors (we want our project in the official distribution),
or the reason proposed by FSF (legal enforcement). This makes the copyright
assignment overly one-sided: it misrepresents the intent of the contributors,
formally purporting they sold their work for one dollar.

**Comment by dgutov** [comment]
@antalk2 You understand you're currently arguing over your rights for 20 lines of comments in a couple of files, right?

> any amount of these outside the agreement is probably considered irrelevant by lawyers

What lawyers? Were you going to try to sue the FSF if this package was never accepted into ELPA?

**Comment by antalk2** [comment]
No, I was not. But major contributors going through all the legality
might be tempted if they find they gave away their rights for nothing.
Or they just might feel they were cheated, but unable to sue, because they
happily "sold" their work without even documenting the actual reason they gave it away.

**Comment by dgutov** [comment]
The contributors "give it away" for FSF to be able to protect the projects better. That is all.

**Comment by wasamasa** [comment]
@antalk2 Regarding the phrase "For one dollar", this is lawyer speak, the reason for it being that you must name some monetary value the copyright assignment is exchanged for. There is no deeper meaning to it.

**Comment by skangas** [comment]
Ping @llasram, any updates on the assignment process? Thanks.

**Comment by wasamasa** [comment]
Given how long this has been going on, I wonder whether inclusion into NonGNU ELPA makes more sense:

- Enabled out of the box in Emacs 28.1+
- No chasing after copyright assignments
- No copyright assignment necessary for contributors

**Comment by skangas** [comment]
> Given how long this has been going on, I wonder whether inclusion into NonGNU ELPA makes more sense:

I believe it is already on NonGNU ELPA: http://elpa.nongnu.org/nongnu/yaml-mode.html

>     * Enabled out of the box in Emacs 28.1+
> 
>     * No chasing after copyright assignments
> 
>     * No copyright assignment necessary for contributors

Also, unfortunately:

   * No chance of being included in GNU Emacs by default. (Personally, I think something as ubiquitous as yaml should have good support OOTB.)

Ideally, we would be able to add it to GNU ELPA. I guess we would only need to procure the assignment from @llasram. If @antalk2 still does not want to sign perhaps his/her contributions could be reverted.

**Comment by llasram** [comment]
Hey, I still have no problem with doing the assignment.  I think last time I looked at it, I ran into a GPG issue (key with lost private key on keyserver? something like that), then the last time someone pinged on the thread I had COVID, and I just failed to do anything to ensure I followed up.

I'll look into it again this weekend, and follow up here if I'm having any difficulties this time around.

Apologies for the extreme delay!

**Comment by llasram** [comment]
Ok, I have (PGP-)signed and sent back the assignment paperwork I was sent back in 2018.  Assuming nothing has changed which requires different paperwork / completing a different process, then I'm done.  I'll comment again once I have confirmation from the FSF.

**Comment by wasamasa** [comment]
The discussion in #104 reminded me that this issue may be obsolete as yaml-mode is part of NonGNU ELPA, which has been added to the `package-archives` as of Emacs 28.1. It can be used from older Emacs versions by customizing `package-archives`, too.

According to the [2022 Emacs survey](https://emacsconf.org/2022/talks/survey/), more than 96% of the correspondents were using Emacs 27.1 or newer. If this trend continues and the 2024/... version of the survey shows the same for Emacs 28.1, then this issue can be closed and the requirement of contributors having assigned copyright to Emacs can be dropped.

To the participants of this thread, what do you think? Does NonGNU ELPA and Emacs 28.1 satisfy the requirement of being able to install yaml-mode out of the box for you?

**Comment by skangas** [comment]
> To the participants of this thread, what do you think? Does NonGNU ELPA and Emacs 28.1 satisfy the requirement of being able to install yaml-mode out of the box for you?

The main drawback I see is that we can't bundle `yaml-mode.el` together with Emacs itself. Arguably, any text-editor worth its salt in 2023 should support YAML out of the box.

On the other hand, we now have `yaml-ts-mode.el` built-in, although that is obviously predicated on an Emacs binary built with treesitter support. I'm also not sure how it compares feature-wise to `yaml-mode.el`. Since it's younger, I assume that some things are still missing (help welcome, BTW).

Ideally, we would could ship Emacs with both the treesitter and non-treesitter modes.

**Comment by yestalgia** [comment]
> To the participants of this thread, what do you think? Does NonGNU ELPA and Emacs 28.1 satisfy the requirement of being able to install yaml-mode out of the box for you?

Late to the party, but since so many contributors have already gone through the FSF assignment process, wouldn’t it make sense to follow through? Inclusion in GNU ELPA makes installation easy for newbies as they don't need to modify package-archives.

Given that most of the groundwork is already done, I reckon go for it.

Thanks for all your efforts.
